[NEW] Add "user is typing" view when some user is typing

macabeus opened the pull request:
@RocketChat/ios

Closes #251 
Closes #690 

- [x] Show "user is typing"
- [x] When the user typing in iOS, [must send the notify on the room](https://rocket.chat/docs/developer-guides/realtime-api/method-calls/notify-room-stream#typing)
- [x] When user change the channel, don't show "user is typing" of previous channel

rafaelks commented:
@macabeus Thanks for your PR!. You can remove a subscription from our socket by calling this method in our web socket:

```
{"msg": "unsub", "id": "subscriptionId"}
````

macabeus commented:
To continue the development of this feature, firstly needs resolve this another issue: #690
We needs a way to unsubscribe the socket when change the channel.

macabeus commented:
In my last commit, I added the feature which allows the socket to unsubscribe from one channel. Then, I fixed two bugs:

- it no longer receives the messages from the previous channel (#690)
- it no longer shows "user is typing" from the previous channel

> Sorry for having broken the atomicity in commits, but I can't really split in minor parts.

rafaelks commented:
@macabeus Feature is looking very good! I've just one adjustment... when the app shows the "User is typing" view, can you adjust the content offset of the collection view, so we don't block the message behind it?

![2017-09-23 08 43 52](https://user-images.githubusercontent.com/551004/30772873-7ef8e5ec-a03b-11e7-8cc3-0ef7fbb53824.gif)

Thank you!

rafaelks commented:
Looking great @macabeus, thank you! Waiting for tests to merge.

macabeus commented:
@rafaelks Fix!

![](https://media.giphy.com/media/xT9IghxRSKm0cgwGfS/giphy.gif)

I tried using the `collectionView?.slk_isAtBottom` to check if the user is at bottom of log, but, this function don't work, then, I created a new function to check if the user is at bottom. Maybe this new function can be used to hidden/show the `buttonScrollToBottom`?

rafaelks commented on Rocket.Chat/Controllers/Chat/ChatViewController.swift:
```diff
@@ -261,8 +265,15 @@ final class ChatViewController: SLKTextViewController {
 
-    // MARK: Message
+    override func textViewDidChange(_ textView: UITextView) {
+        if textView.text.characters.count > 0 {
```
You **could** use `isEmpty` here.

rafaelks commented:
@macabeus Yes, you can use the same function. Can you do that on this PR, please?

macabeus commented:
I fixed the bug in `buttonScrollToBottom` without having to use my code, then, I made another pull request: #721
And, the function that I created for this pullrequest isn't necessary in `buttonScrollToBottom`.
